File: hash-table/make-sum-divisible-by-p/make-sum-divisible-by-p.js
/**
 * @param {number[]} nums
 * @param {number} p
 * @return {number}
 */
function minSubarray(nums, p) {
  const target = (nums.reduce((acc, curr) => acc + curr, 0)) % p;

  if (!target) {
    return 0;
  }

  const remains = {};
  remains[0] = -1;

  let res = nums.length;
  let sum = 0;
  for (let j = 0; j < nums.length; j++) {
    sum += nums[j];
    const currentRemain = sum % p;
    let key = (currentRemain - target)%p;
    key = key < 0 ? key + p : key;
    const i = remains[key];

    if (i !== undefined) {
      res = Math.min(j - i, res);
    }

    remains[currentRemain] = j;
  }

  return res === nums.length ? -1 : res;
}

File: hash-table/make-sum-divisible-by-p/make-sum-divisible-by-p.go
package make

func minSubarray(nums []int, p int) int {
	sum := 0
	for _, num := range nums {
		sum += num
	}

	target := positiveRemain(sum, p)

	if target == 0 {
		return 0
	}

	lookup := make(map[int]int)
	lookup[0] = -1
	length := len(nums)
	sum = 0
	for j, num := range nums {
		sum += num
		remain := positiveRemain(sum, p)
		key := positiveRemain(remain-target, p)

		if i, ok := lookup[key]; ok {
			currentLength := j - i
			if currentLength < length {
				length = currentLength
			}
		}

		lookup[remain] = j
	}

	if length == len(nums) {
		return -1
	}

	return length

}

func positiveRemain(num, div int) int {
	remain := num % div
	if num < 0 {
		return remain + div
	}
	return remain
}
